File: exercise_01.c
#include <pthread.h>
#include <stdio.h>

int i = 0;

// Note the return type: void*
void* incrementingThreadFunction() {
	// TODO: increment i 1_000_000 times
	for (int c = 0; c < 1000000; c = c++) {
		i = i++;
	}
	return NULL;
}

void* decrementingThreadFunction() {
	// TODO: decrement i 1_000_000 times
	for (int c = 0; c < 1000000; c = c++) {
		i = i--;
	}
	return NULL;
}


int main() {
	// TODO: declare incrementingThread and decrementingThread (hint: google pthread_create)
	pthread_t thread_1;
	pthread_create(&incrementingThread, NULL, incrementingThreadFunction, NULL);
	pthread_t thread_2;
	pthread_create(&decrementingThread, NULL, decrementingThreadFunction, NULL);

	pthread_join(incrementingThread, NULL);
	pthread_join(decrementingThread, NULL);

	printf("The magic number is: %d\n", i);
	return 0;
}

File: exercise_01.go
// Use `go run foo.go` to run your program

package main

import (
    . "fmt"
    "runtime"
    "time"
)

var i = 0

func incrementing() {
    //TODO: increment i 1000000 times
	for c:=0; c<1000000; c++{
		i++
	}
}

func decrementing() {
    //TODO: decrement i 1000000 times
	for c:=0; c<1000000; c++{
		i--
	}
}

func main() {
    runtime.GOMAXPROCS(runtime.NumCPU())    // I guess this is a hint to what GOMAXPROCS does...
	                                    // Try doing the exercise both with and without it!

    // TODO: Spawn both functions as goroutines
	go incrementing()
	time.Sleep(100*time.Millisecond)
	go decrementing()
	
    // We have no way to wait for the completion of a goroutine (without additional syncronization of some sort)
    // We'll come back to using channels in Exercise 2. For now: Sleep.
    time.Sleep(100*time.Millisecond)
	go decrementing()
    Println("The magic number is:", i)
}